src/util.c

```diff
@@ -18,17 +18,21 @@ writable_buffer_ensure_space(writable_buffer *wb, uint64 minspace)
       minspace = 2 * wb->buffer_capacity;
    }
 
-   void *oldptr  = wb->can_free ? wb->buffer : NULL;
-   void *newdata = platform_realloc(
-      wb->heap_id, writable_buffer_length(wb), oldptr, minspace);
+   void *newdata = NULL;
+   if (wb->can_free) {
+      newdata = platform_realloc(
+         wb->heap_id, wb->buffer_capacity, wb->buffer, minspace);
```

Comment:
@rtjohnso - Do note there is still a peculiarity here, even though we are using `wb->buffer_capacity` as old-length passed to realloc().

This old-length value (i.e. `wb->buffer_capacity`) will only be used to copy-over data after reallocating new memory.

Your reasoning was that for `free(ptr)`, we should pass-in the allocated memory-size (i.e. buffer-capacity) and not the actual "in-use" data length.

The pair of `(wb->buffer, wb->buffer_capacity)` is not sufficient for `free()` to correctly identify this memory fragment. This is because allocation could have added pad-bytes at the time of allocation, rounding-up requested 'capacity' to something larger (for cache-alignment).

So the argument that `free()` will need to use the `buffer-capacity` rather than `buffer-length` to locate the fragment-to-be-freed is also not kosher.